AXNotification enum ought to be an X macro

rkirsling opened the pull request:
#### 129ce14b05a4a086cd7ddb915f9ad64c116873aa
<pre>
AXNotification enum ought to be an X macro
<a href="https://bugs.webkit.org/show_bug.cgi?id=270893">https://bugs.webkit.org/show_bug.cgi?id=270893</a>

Reviewed by Darin Adler.

AXNotification is a giant and frequently updated enum.
Every time it is updated, exhaustive switch statements must be updated accordingly.

This is a very appropriate place for an X macro --
selfishly, it will help PlayStation port avoid build breaks (since we handle this enum in WKAPICast),
but irrespective of platform, it also means that AXLogger won&apos;t need to be updated each time AXNotification is updated.

* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.h:

Canonical link: <a href="https://commits.webkit.org/276393@main">https://commits.webkit.org/276393@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f65af2aa755a41198b1f2e48a3c02647f23c5999

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44538 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47189 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40541 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27633 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21005 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45114 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20670 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38332 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17697 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39476 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2585 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40731 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39752 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48812 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19498 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16031 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43561 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20840 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42308 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9908 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21168 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20496 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -139,6 +139,110 @@ class AccessibilityReplacedText {
 };
 
+#define WEBCORE_AXNOTIFICATION_KEYS_DEFAULT(macro) \
+    macro(AccessKeyChanged) \
```
This looks strictly worse. I don't understand why we want to adopt this change.

rkirsling replied:
I linked the explanation above but I can link it again here: https://github.com/WebKit/WebKit/pull/23035#discussion_r1494036159

Introducing `WebKit/UIProcess/API/C/WKAXTypes.h` with enums like...
```cpp
enum AXNotification {
     // must be same as WebCore::AXObjectCache::AXNotification
     kWKAXAccessKeyChanged,
     ...
 };
 typedef uint32_t WKAXNotification;
 ```
 ...and extending `WebKit/UIProcess/API/C/WKAPICast.h` with casts like...
 ```cpp
 inline WKAXNotification toAPI(WebCore::AXObjectCache::AXNotification notification)
 {
     switch (notification) {
     case WebCore::AXObjectCache::AXNotification::AXAccessKeyChanged:
         return kWKAXAccessKeyChanged;
     ...
     }
 }
```
...feels rather hard to justify if only one platform is planning to use these casts.

Since the idea is just to ensure that the API casts get updated whenever the WebCore enums get updated, the ideal situation would be to have this happen automatically, without requiring developers to be conscious of it. This automation is achieved by [X macros](https://en.wikipedia.org/wiki/X_macro), which allow us to do an exhaustive `switch` on a enum without being concerned with the specific values of that enum changing. Indeed, forgetting about any platform-specific needs, in AXLogger, we're able to reduce a `switch` from 286 lines to just 8, and those 8 lines needn't be touched the next time AXNotification changes.

I understand that while this is commonplace in JSC (looks like we may have almost 200 of them), it surely feels quite foreign in WebCore. But I'm not sure of a different tool that fits this problem so well.

rkirsling replied:
Seems like I've overestimated how foreign this actually is outside of JSC. We have X macros defined in...
- [`Source/WebCore/bindings/js/WebCoreBuiltinNames.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/bindings/js/WebCoreBuiltinNames.h#L38-L740)
- [`Source/WebCore/cssjit/SelectorCompiler.cpp`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/cssjit/SelectorCompiler.cpp#L85-L150)
- [`Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp#L137-L155)
- [`Source/WebCore/inspector/InspectorCanvasCallTracer.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/inspector/InspectorCanvasCallTracer.h#L84-L190)
- [`Source/WebCore/inspector/InstrumentingAgents.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/inspector/InstrumentingAgents.h#L110-L138)
- [`Source/WebCore/page/ResourceUsageData.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/page/ResourceUsageData.h#L37-L58)
- [`Source/WebCore/platform/CommonAtomStrings.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/CommonAtomStrings.h#L33-L77)
- [`Source/WebCore/platform/Logging.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/Logging.h#L39-L151)
- [`Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h#L81-L107)
- [`Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.h#L35-L74)
- [`Source/WebCore/PAL/pal/Logging.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/PAL/pal/Logging.h#L40-L48)
- [`Source/WebKit/Platform/Logging.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebKit/Platform/Logging.h#L41-L133)
- [`Source/WebKit/UIProcess/ios/WKContentViewInteraction.h`](https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/ios/WKContentViewInteraction.h#L157-L236)
- [`Source/WTF/wtf/Logging.h`](https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/Logging.h#L39-L52)
- [`Source/WTF/wtf/SystemTracing.h`](https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/SystemTracing.h#L219-L227)
- [`Tools/TestRunnerShared/EventSerialization/mac/EventSerializerMac.mm`](https://github.com/WebKit/WebKit/blob/main/Tools/TestRunnerShared/EventSerialization/mac/EventSerializerMac.mm#L87-L152)
- [`Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp`](https://github.com/WebKit/WebKit/blob/main/Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp#L107-L125)

...and a few more files which are GTK/WPE-specific.

So I revoke my hedging; this is not only the right tool for the job, it's also a perfectly normal thing for us to do.

darinadler commented on Source/WebCore/accessibility/AXLogger.cpp:
```diff
@@ -543,290 +543,12 @@ TextStream& operator<<(TextStream& stream, AXRelationType relationType)
-        stream << "AXTextCompositionEnded";
+#define WEBCORE_LOG_AXNOTIFICATION(name) \
+    case AXObjectCache::AXNotification::name: \
+        stream << #name; \
```
If we put `“AX” #name` here then we could leave all the repeated AX out of the macro arguments.

rkirsling replied:
Great idea!

darinadler commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -355,102 +459,9 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
-        AXDraggingEnteredDropZone,
-        AXDraggingDropped,
-        AXDraggingExitedDropZone
+#define WEBCORE_DEFINE_AXNOTIFICATION_ENUM(name) name,
```
And here `AX##name`.

rkirsling committed:
AXNotification enum ought to be an X macro
https://bugs.webkit.org/show_bug.cgi?id=270893

Reviewed by Darin Adler.

AXNotification is a giant and frequently updated enum.
Every time it is updated, exhaustive switch statements must be updated accordingly.

This is a very appropriate place for an X macro --
selfishly, it will help PlayStation port avoid build breaks (since we handle this enum in WKAPICast),
but irrespective of platform, it also means that AXLogger won't need to be updated each time AXNotification is updated.

* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.h:

Canonical link: https://commits.webkit.org/276393@main
